### PR TITLE
recipe: add version 6.1.1 for pytest

### PR DESCRIPTION
**Pull request by Bidibulke**
add recipe for new dependency py-iniconfig
recipe: add version 6.1.1 for pytest

add recipe for new dependency py-iniconfig

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-iniconfig/package.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    * maintains order of sections and entries
+    * supports multi-line values with or without line-continuations
+    * supports '#' comments everywhere
+    * raises errors with proper line-numbers
+    * no bells and whistles like automatic substitutions
+    * iniconfig raises an Error if two sections have the same name.
+    """
+
+    homepage = "https://github.com/RonnyPfannschmidt/iniconfig"
+    url      = "https://pypi.io/packages/source/i/iniconfig/iniconfig-1.1.1.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-iniconfig/package.py]: https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py]: Undo this change

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 3]
```diff
@@ -10,8 +10,11 @@ class PyPytest(PythonPackage):
     """pytest: simple powerful testing with Python."""
 
     homepage = "http://pytest.org/"
-    pypi = "pytest/pytest-5.2.1.tar.gz"
+    url      = "https://pypi.io/packages/source/p/pytest/pytest-5.2.1.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py]: Undo this change

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 4]
```diff
@@ -10,8 +10,11 @@ class PyPytest(PythonPackage):
...
 
+    import_modules = ['pytest']
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py]: Undo this change

File: var/spack/repos/builtin/packages/py-iniconfig/package.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    iniconfig is a small and simple INI-file parser module having a unique
+    set of features:
+
+    * tested against Python2.4 across to Python3.2, Jython, PyPy
+    * maintains order of sections and entries
+    * supports multi-line values with or without line-continuations
+    * supports '#' comments everywhere
+    * raises errors with proper line-numbers
+    * no bells and whistles like automatic substitutions
+    * iniconfig raises an Error if two sections have the same name.
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-iniconfig/package.py]: `spack info` removes newline characters, so this likely won't be formatted like you expect. I would just keep the first line description, that's probably enough.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 6]
```diff
@@ -37,7 +38,8 @@ class PyPytest(PythonPackage):
     depends_on('py-setuptools@30.3:', when='@3.9.0:3.9.1', type=('build', 'run'))
     depends_on('py-setuptools', type=('build', 'run'))
     depends_on('py-setuptools-scm', when='@3.1:', type='build')
-    depends_on('py-py@1.5.0:', when='@3.3:', type=('build', 'run'))
+    depends_on('py-py@1.8.2:', when='@:6.0.0', type=('build', 'run'))
+    depends_on('py-py@1.5.0:', when='@3.3:6.0', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py]:
```suggestion
    depends_on('py-py@1.8.2:', when='@6:', type=('build', 'run'))
    depends_on('py-py@1.5.0:', when='@3.3:5.999', type=('build', 'run'))
```
Is this what you're trying to do here?

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 7]
```diff
@@ -62,3 +64,5 @@ class PyPytest(PythonPackage):
     depends_on('py-importlib-metadata@0.12:', when='@4.6:5.0', type=('build', 'run'))
     depends_on('py-importlib-metadata@0.12:', when='@5.1: ^python@:3.7', type=('build', 'run'))
     depends_on('py-wcwidth', when='@4.5:', type=('build', 'run'))
+    depends_on('py-iniconfig', when='@5.0:', type=('build', 'run'))
+    depends_on('py-toml', when='@5.0:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py]: Do you mean `@6:`?

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 8]
```diff
@@ -62,3 +64,5 @@ class PyPytest(PythonPackage):
     depends_on('py-importlib-metadata@0.12:', when='@4.6:5.0', type=('build', 'run'))
     depends_on('py-importlib-metadata@0.12:', when='@5.1: ^python@:3.7', type=('build', 'run'))
     depends_on('py-wcwidth', when='@4.5:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py]: This dep doesn't seem to be needed anymore

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-pytest/package.py (hunk 6 above)
Comment (Bidibulke) [reply var/spack/repos/builtin/packages/py-pytest/package.py]: yes. Changes done.

File: var/spack/repos/builtin/packages/py-pytest/package.py (hunk 7 above)
Comment (Bidibulke) [reply var/spack/repos/builtin/packages/py-pytest/package.py]: yes again. Done.

File: var/spack/repos/builtin/packages/py-pytest/package.py (hunk 8 above)
Comment (Bidibulke) [reply var/spack/repos/builtin/packages/py-pytest/package.py]: * I added a version constraint for `py-wcwith`
* `py-importlib-metada` is still neeeded if python < 3.8 so I changed the constraint to `python@:3.7.999`

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 9]
```diff
@@ -60,5 +62,7 @@ class PyPytest(PythonPackage):
     depends_on('py-pathlib2@2.2.0:', when='@3.7.1: ^python@:3.5', type=('build', 'run'))
     depends_on('py-pathlib2', when='@3.7.0 ^python@:3.5', type=('build', 'run'))
     depends_on('py-importlib-metadata@0.12:', when='@4.6:5.0', type=('build', 'run'))
-    depends_on('py-importlib-metadata@0.12:', when='@5.1: ^python@:3.7', type=('build', 'run'))
-    depends_on('py-wcwidth', when='@4.5:', type=('build', 'run'))
+    depends_on('py-importlib-metadata@0.12:', when='@5.1:5.999 ^python@:3.7.999', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py]: This dependency is still needed in newer versions, you can undo this change.

Comment (Bidibulke) [reply var/spack/repos/builtin/packages/py-pytest/package.py]: done

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 10]
```diff
@@ -62,3 +64,6 @@ class PyPytest(PythonPackage):
     depends_on('py-importlib-metadata@0.12:', when='@4.6:5.0', type=('build', 'run'))
     depends_on('py-importlib-metadata@0.12:', when='@5.1: ^python@:3.7', type=('build', 'run'))
     depends_on('py-wcwidth', when='@4.5:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py]: I think you need to get rid of this line, right? Otherwise you'll still depend on wcwidth in the latest version.

Comment (Bidibulke) [reply var/spack/repos/builtin/packages/py-pytest/package.py]: Oops, of course. done

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 11]
```diff
@@ -60,5 +62,6 @@ class PyPytest(PythonPackage):
     depends_on('py-pathlib2@2.2.0:', when='@3.7.1: ^python@:3.5', type=('build', 'run'))
     depends_on('py-pathlib2', when='@3.7.0 ^python@:3.5', type=('build', 'run'))
     depends_on('py-importlib-metadata@0.12:', when='@4.6:5.0', type=('build', 'run'))
-    depends_on('py-importlib-metadata@0.12:', when='@5.1: ^python@:3.7', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py:63]: You need to undo this change again

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-pytest/package.py (hunk 11 above)
Comment (Bidibulke) [reply var/spack/repos/builtin/packages/py-pytest/package.py:63]: done.

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 12]
```diff
@@ -60,5 +62,7 @@ class PyPytest(PythonPackage):
...
-    depends_on('py-wcwidth', when='@4.5:', type=('build', 'run'))
+    depends_on('py-importlib-metadata@0.12:', when='@5.1: ^python@3.7', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py]: You're missing a colon.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-pytest/package.py (hunk 12 above)
Comment (Bidibulke) [reply var/spack/repos/builtin/packages/py-pytest/package.py]: sorry. I hope this time every thing will be ok.

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
@trws seeing the same `mypy` failure on this PR as well

**Comment by Bidibulke** [comment]
I try `spack style` (with `mypy` 0.800 instead of 0.740) on every modified files and I don't have any errors, but when I run `spack style`, I still have an error but I don't know where. I also try `spack style -a`, there are some errors but not on my modified files.
If you have any idea to fix the problem, I'll be glad to have it.

In the same time, I added a new version of `py-pytest` (6.2.1) and `py-py` (1.8.2 and 1.9.0). There was a mistake in `py-pytest` recipe where I ask for `py-py` 1.8.2, but I didn't change the recipe of `py-py`.

**Comment by adamjstewart** [comment]
The `spack style` issue was fixed in develop, hence why the tests pass now.

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 13]
```diff
@@ -37,17 +40,20 @@ class PyPytest(PythonPackage):
     depends_on('py-setuptools@30.3:', when='@3.9.0:3.9.1', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py:41]: Setuptools 42+ is needed now

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 14]
```diff
@@ -37,17 +40,20 @@ class PyPytest(PythonPackage):
...
     depends_on('py-setuptools', type=('build', 'run'))
     depends_on('py-setuptools-scm', when='@3.1:', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py:44]: Setuptools-scm 3.4+ is needed now

File: var/spack/repos/builtin/packages/py-pytest/package.py [hunk 15]
```diff
@@ -37,17 +40,20 @@ class PyPytest(PythonPackage):
...
+    depends_on('py-py@1.5.0:', when='@3.3:5.999', type=('build', 'run'))
     depends_on('py-py@1.4.33:', when='@3.1.2:3.2.3,3.2.5:3.2.999', type=('build', 'run'))
     depends_on('py-py@1.4.33:1.4.999', when='@3.2.4', type=('build', 'run'))
     depends_on('py-py@1.4.29:', when='@:3.1.1', type=('build', 'run'))
     depends_on('py-six@1.10.0:', when='@3.3:4', type=('build', 'run'))
     depends_on('py-packaging', when='@4.6:', type=('build', 'run'))
+    depends_on('py-attrs@19.2.0:', when='@6.1.1:', type=('build', 'run'))
     depends_on('py-attrs@17.2.0:', when='@3.3:3.4', type=('build', 'run'))
     depends_on('py-attrs@17.4.0:', when='@3.5:', type=('build', 'run'))
     depends_on('py-more-itertools@4.0.0:', when='@3.5.1:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pytest/package.py]: This dep isn't needed in the latest version

**Review by adamjstewart — CHANGES_REQUESTED** [review]
It's easier to review PRs when there is only a single package per PR 😄

File: var/spack/repos/builtin/packages/py-pytest/package.py (hunk 15 above)
Comment (Bidibulke) [reply var/spack/repos/builtin/packages/py-pytest/package.py]: I kept the line because in `pytest-6.2.2/setup.cfg` file, there is still a line 
```
attrs>=19.2.0
```

I modified the *when* for `py-attrs@19.20` et `py-attrs@17.4.0`

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pytest/package.py]: I'm commenting on more-itertools, not attrs

Comment (Bidibulke) [reply var/spack/repos/builtin/packages/py-pytest/package.py]: sorry, I was focus on the modified line.

**Review by adamjstewart — APPROVED** [review]
